sdk/openai/azure-ai-openai-assistants/src/main/java/com/azure/ai/openai/assistants/implementation/streaming/OpenAIServerSentEvents.java

```diff
@@ -0,0 +1,157 @@
+package com.azure.ai.openai.assistants.implementation.streaming;
+
+import com.azure.ai.openai.assistants.implementation.models.AssistantStreamEvent;
+import com.azure.ai.openai.assistants.models.StreamUpdate;
+import com.azure.core.util.BinaryData;
+import reactor.core.publisher.Flux;
+import reactor.core.scheduler.Schedulers;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.azure.ai.openai.assistants.implementation.models.AssistantStreamEvent.DONE;
+import static com.azure.ai.openai.assistants.implementation.models.AssistantStreamEvent.ERROR;
+
+public final class OpenAIServerSentEvents {
+
+    // Server sent events are divided by 2 line breaks, therefore we need to account for 4 bytes containing 2 CRLF characters.
+    private static final int LINE_BREAK_CHAR_COUNT_THRESHOLD = 2;
+
+    private final StreamTypeFactory eventDeserializer = new StreamTypeFactory();
+    private final Flux<ByteBuffer> source;
+    private ByteArrayOutputStream outStream;
+
+    public OpenAIServerSentEvents(Flux<ByteBuffer> source) {
+        this.source = source;
+        this.outStream = new ByteArrayOutputStream();
+    }
+
+    public Flux<StreamUpdate> getEvents() {
+        return mapEventStream();
+    }
+
+    /**
+     * Maps the byte buffer to a stream of server sent events.
+     *
+     * @return A stream of server sent events deserialized into StreamUpdates.
+     */
+    private Flux<StreamUpdate> mapEventStream() {
+        return source
+            .publishOn(Schedulers.boundedElastic())
+            .concatMap(byteBuffer -> {
+                List<StreamUpdate> values = new ArrayList<>();
+                byte[] byteArray = byteBuffer.array();
+                // We check whether we ended the last byteBuffer with a line feed or not, in case we need to close this
+                // chunk immediately
+                byte[] outByteArray = outStream.toByteArray();
+                int lineBreakCharsEncountered = outByteArray.length > 0 && isByteLineFeed(outByteArray[outByteArray.length  - 1]) ? 1 : 0;
+
+                for (byte currentByte : byteArray) {
+                    outStream.write(currentByte);
+                    if (isByteLineFeed(currentByte)) {
+                        lineBreakCharsEncountered++;
+
+                        // We are looking for 2 line breaks to signify the end of a server sent event.
+                        if (lineBreakCharsEncountered == LINE_BREAK_CHAR_COUNT_THRESHOLD) {
```

Comment:
Suggestion: "threshold" -> "marker"/"sentinel"/etc. (threshold makes me think of a scalar that can be exceeded or "crossed")

Reply:
Thank you, now that you mention it, threshold does ring like a "minimum value that needs to be met" which doesn't reflect the reality here that we need exactly `2`.